Playerstats command

Floweynt opened the pull request:
Implement the `/playerstats` command. 

- make API wrappers async & thread-safe
- implement playerstats command for 1.8.9
- bundle brigadier client-side (on 1.8.9) to hopefully enable better abstraction cross-version 
- move class `LevelHead` -> module `common` since it doesn't depend on MC.

TODO:
- [x] split `HypixelAbstractionLayer` to two parts: one to handle API and the other to handle game/transient state
- [ ] configurable cache size
- [x] port to other versions
- [x] clean up rate-limit logic
- [x] implement for other games
- [x] add pretty colors

Non goals (for this PR):
- Overlay

Commit:
Implement stats command & off-main async API support

Floweynt commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/HypixelAbstractionLayer.java:
```diff
@@ -22,150 +22,181 @@
+	private final CachedAPI<String, Integer> skywardsExpApi = createLevel(RequestDataType.SKYWARS_EXPERIENCE);
+
+	@SneakyThrows // propagate interrupted exception
+	public void shutdown() {
```
So it seems like I don't need to explicitly clean up the thread, I never call this but Minecraft seems to stop cleanly. Removing this would simplify the logic in the worker thread.

moehreag replied:
It would probably be best to make use of the `ForkJoinPool` held by `ThreadExecutor` which is also used for other requests

moehreag commented on 1.21.4/src/main/java/io/github/axolotlclient/api/chat/ChannelInvitesScreen.java:
```diff
@@ -120,7 +120,7 @@ public InvitesListEntry(ChannelInvite invite) {
 			public void render(GuiGraphics graphics, int index, int top, int left, int width, int height, int mouseX, int mouseY, boolean hovering, float partialTick) {
 				graphics.drawString(font, Component.translatable("api.channels.invite.name", invite.channelName()), left + 2, top + 2, -1);
-				graphics.drawString(font, Component.translatable("api.channels.invite.from", UUIDHelper.getUsername(invite.fromUuid())).withStyle(Style.EMPTY.withItalic(true)), left + 15, top + height - font.lineHeight - 1, 0x808080);
+				graphics.drawString(font, Component.translatable("api.channels.invite.from", UUIDHelper.UUID_TO_USERNAME.getBlocking(invite.fromUuid())).withStyle(Style.EMPTY.withItalic(true)), left + 15, top + height - font.lineHeight - 1, 0x808080);
```
What is the reason for this specific method using the cache directly instead of the `tryGet` methods?

moehreag commented on 1.21.4/src/main/java/io/github/axolotlclient/mixin/PlayerListHudMixin.java:
```diff
@@ -189,8 +189,10 @@ public abstract class PlayerListHudMixin {
 
-			render = String.valueOf(HypixelAbstractionLayer.getPlayerLevel(
-				playerListEntry2.getProfile().getId().toString().replace("-", ""), LevelHeadMode.BEDWARS));
+			String uuid = playerListEntry2.getProfile().getId().toString().replace("-", "");
```
Not on you, but this should probably make use of `UUIDHelper#toUndashed`

moehreag commented on 1.21.4/src/main/java/io/github/axolotlclient/modules/screenshotUtils/ImageScreen.java:
```diff
@@ -91,7 +91,7 @@ protected void init() {
 		if (image instanceof ImageInstance.Remote remote) {
 			layout.setHeaderHeight(38);
-			header.addChild(new StringWidget(Component.translatable("gallery.image.upload_details", UUIDHelper.getUsername(remote.uploader()), remote.sharedAt().atZone(ZoneId.systemDefault()).format(AxolotlClientCommon.getInstance().formatter)), font));
+			header.addChild(new StringWidget(Component.translatable("gallery.image.upload_details", UUIDHelper.UUID_TO_USERNAME.getBlocking(remote.uploader()), remote.sharedAt().atZone(ZoneId.systemDefault()).format(AxolotlClientCommon.getInstance().formatter)), font));
```
This should also use the `tryGet` methods (Minecraft's `Component`/`Text` just call `toString` on the arguments).


![image](https://github.com/user-attachments/assets/2e8214e9-b608-41cf-be6e-280c2672361c)

moehreag commented on 1.21/src/main/java/io/github/axolotlclient/mixin/PlayerListHudMixin.java:
```diff
@@ -187,9 +187,10 @@ public abstract class PlayerListHudMixin {
-			render = String.valueOf(HypixelAbstractionLayer.getPlayerLevel(playerListEntry2
-					.getProfile().getId().toString().replace("-", ""),
-				LevelHeadMode.BEDWARS));
+			String uuid = playerListEntry2.getProfile().getId().toString().replace("-", "");
```
Again not on you, but this should probably make use of `UUIDHelper#toUndashed` as well

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/HypixelAbstractionLayer.java:
```diff
@@ -22,150 +22,181 @@
-	}
+	// TODO: Someone who's better at async algorithms should port this to a scheduled executor service. This implementation is certainly janky and VERY VERY BAD!
+	// TODO: after a request fails {n} times, we should give up...
+	private final Thread worker = new Thread("HypixelAbstractionLayerWorker") {
```
I'm not sure if assigning just a single worker thread to this is a good idea, since usually requests happen in bursts after you switch worlds. The backend tries its best to not ratelimit requests so it shouldn't be an issue to allow more requests to take place concurrently. It's only a bit complicated since if the backend runs out of API requests, the client shouldn't spam the same requests over and over but instead have some sort of cooldown.

Floweynt replied:
The thread only exists for starting requests and handling rate limiting. It dispatches network stuff to whatever the `HttpClient` (which uses the common fork join pool iirc). I don't want to spam the backend, so I have a relatively conservative algo. Technically, we can get rid of this thread completely if we come up with a better technique for handling rate limits.

Floweynt replied:
This has been rewritten

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/HypixelAbstractionLayer.java:
```diff
@@ -22,150 +22,181 @@
+	private final CachedAPI<String, Integer> bedwarsLevelApi = createLevel(RequestDataType.BEDWARS_LEVEL);
+
+	@Getter
+	private final CachedAPI<String, Integer> skywardsExpApi = createLevel(RequestDataType.SKYWARS_EXPERIENCE);
```
should probably be `skywarsLevelApi`, using `ExpCalculator` to calculate the level from the experience (as it's used currently)

moehreag commented on common/src/main/java/io/github/axolotlclient/modules/hypixel/levelhead/LevelHeadMode.java:
```diff
@@ -22,6 +22,17 @@
-	NETWORK, BEDWARS, SKYWARS
+	NETWORK(HypixelAbstractionLayer.getInstance().getNetworkLevelApi()),
+	BEDWARS(HypixelAbstractionLayer.getInstance().getBedwarsLevelApi()),
+	SKYWARS( HypixelAbstractionLayer.getInstance().getSkywardsExpApi());
```
There shouldn't be a space here

moehreag commented on common/src/main/java/io/github/axolotlclient/util/CachedAPI.java:
```diff
@@ -0,0 +1,75 @@
+
+    private final LoadingCache<K, CompletableFuture<V>> cache = CacheBuilder
+        .newBuilder()
+        .maximumSize(100)
```
It might make sense to allow different instances of `CachedAPI` to have different maximum sizes since f.e. the level data will usually reach multiple hundred players quite quickly

moehreag reviewed (commented):
Generally a lot of good things, however there's a few issues as well.
*If possible please also don't change the import order*

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/mixin/commands/LocalClientPlayerEntityMixin.java:
```diff
@@ -0,0 +1,46 @@
+		at = @At(value = "HEAD"),
+		cancellable = true
+	)
+	private void handleChatEvents(String string, CallbackInfo ci, @Local(argsOnly = true) LocalRef<String> stringRef) {
```
The captured local is unused

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/commands/ClientCommands.java:
```diff
@@ -0,0 +1,124 @@
+		ClientCommandInfo source = buildClientSource(client);
+		// cancel if present
+		command = command.trim();
+		if (command.startsWith("/")) {
```
This check is unnecessary

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/commands/OfflinePlayerArgument.java:
```diff
@@ -0,0 +1,74 @@
+import net.minecraft.client.network.PlayerInfo;
+import net.minecraft.client.network.handler.ClientPlayNetworkHandler;
+
+public class OfflinePlayerArgument implements ArgumentType<OfflinePlayerArgument.OfflinePlayerInfo> {
```
Why is this class called `OfflinePlayerArgument` if it provides suggestions for all *online* players as well as validation for player names in general?

Floweynt replied:
In retrospect it was a bad choice. It's an offline player in the sense that the player isn't necessarily connected to the server (bukkit terminology).

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/commands/ClientCommands.java:
```diff
@@ -0,0 +1,124 @@
+		Text message = new LiteralText(e.getMessage());
+		String context = e.getContext();
+		return context != null ?
+			new TranslatableText("lcu.command.parse_error", message, e.getCursor(), context) : message;
```
What does `lcu` stand for? And where does it stem from? Is it necessary to add credits for another project to these files?

Floweynt replied:
My own project, no need for credits. I don't have it published, but it's just backporting utils.

moehreag commented on 1.8.9/src/main/java/io/github/axolotlclient/mixin/commands/ChatScreenMixin.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @WrapMethod(method = "setMessageHistory")
+    private void addClientSideSuggestions(String[] suggestions, Operation<Void> original) {
```
I think it would be cool to have suggestions in the style of modern versions as well as I couldn't get them to show, however the autocomplete works nicely.

Floweynt replied:
Probably worth porting over, but it's outside the scope of this PR.

moehreag committed:
Add maximum number of request attempts

- correct winstreak property

moehreag committed:
Fix a few minor issues

moehreag committed:
Add translations for StatsMod

Floweynt commented:
> _If possible please also don't change the import order_

Sorry, IDE configuration issue. Is there a standard import order specification?

moehreag commented:
> > _If possible please also don't change the import order_
> 
> Sorry, IDE configuration issue. Is there a standard import order specification?

We have it set to 
`import javax.*` (with subpackages)
`import java.*` (with subpackages)
_blank_
import all other imports
_blank_
import static all other imports

Ideally this would be configured in the .editorconfig or another project-specific formatting file but since we don't know how to do that that's a bit difficult

moehreag committed:
Port to the other versions

moehreag committed:
Remove extra newline

moehreag committed:
Send command responses on the client thread

moehreag committed:
Actually send command responses on the client thread

moehreag committed:
implement `PlayerData` hypixel request type

moehreag committed:
remove winstreak stat from combined entries

Commit:
some async changes, use player_data api

moehreag committed:
Synchronize StatsMod across versions

moehreag committed:
Correct spacing for rank titles

moehreag committed:
Fix a few introduced bugs, clean up a bit

Commit:
autocomplete ignores case

moehreag committed:
Remove potentially blocking behavior from render methods

moehreag committed:
(1.21.x) Fix ImageScreen uploader names

moehreag committed:
fix a bunch of bugs

- use gson 2.8.0 in the common module since that's the oldest one anyone's going to use. Unfortunately it still doesn't support record deserialization.
- move more code to asynchronous methods

Floweynt commented:
> > > > _If possible please also don't change the import order_
> > > 
> > > 
> > > Sorry, IDE configuration issue. Is there a standard import order specification?
> > 
> > 
> > We have it set to `import javax.*` (with subpackages) `import java.*` (with subpackages) _blank_ import all other imports _blank_ import static all other imports
> > Ideally this would be configured in the .editorconfig or another project-specific formatting file but since we don't know how to do that that's a bit difficult
> 
> That's IntelliJ's default.
> 
> Also I'm pretty sure that cannot be configured with .editorconfig unfortunately.

Spotless can enforce import order (as well as license checks), might be worth setting that up but gradle is annoying.

moehreag committed:
fix BedwarsPlayerStats fake stat generation

Commit:
prettify stats & skywars

Commit:
auth: fix indexOf() returning -1 preventing auth

Commit:
port to other version

moehreag committed:
StatsMod/Bedwars: Also color the star icon

moehreag committed:
StatsMod: Clean up a bit, add missing translation

moehreag committed:
Merge branch 'dev' into pr/139

# Conflicts:
#	1.8.9/src/main/resources/axolotlclient.mixins.json

moehreag committed:
fix ApiResult#error returning the wrong instance

moehreag committed:
Add duels statistics

moehreag committed:
Merge branch 'dev' into pr/139

# Conflicts:
#	common/src/main/resources/assets/axolotlclient/lang/en_us.json

moehreag committed:
fix imports

moehreag committed:
add missing translation

moehreag committed:
format

moehreag commented:
Thanks!
